packages/apollo-link-state/src/__tests__/advanced.ts

```diff
@@ -104,4 +104,67 @@ describe('server and client state', () => {
       client.mutate({ mutation, variables });
     }, 10);
   });
+
+  it('correctly propagates an error from a client-state resolver', async done => {
+    const data = {
+      list: {
+        __typename: 'List',
+        items: [
+          { __typename: 'ListItem', id: 1, name: 'first', isDone: true },
+          { __typename: 'ListItem', id: 2, name: 'second', isDone: false },
+        ],
+      },
+    };
+    // mocked endpoint acting as server data
+    const http = new ApolloLink(() => Observable.of({ data }));
+
+    const local = withClientState({
+      resolvers: {
+        Query: {
+          hasBeenIllegallyTouched: (_, _v, _c) => {
+            throw new Error('Illegal Query Operation Occurred');
+          },
+        },
+
+        Mutation: {
+          touchIllegally: (_, _v, _c) => {
+            throw new Error('Illegal Mutation Operation Occurred');
+          },
+        },
+      },
+    });
+
+    const client = new ApolloClient({
+      link: local.concat(http),
+      cache: new InMemoryCache(),
+    });
+
+    const variables = { id: 1 };
+    const query = gql`
+      query hasBeenIllegallyTouched($id: Int!) {
+        hasBeenIllegallyTouched(id: $id) @client
+      }
+    `;
+    const mutation = gql`
+      mutation SelectItem($id: Int!) {
+        touchIllegally(id: $id) @client
+      }
+    `;
+
+    try {
+      await client.query({ query, variables });
+      done.fail('Should have thrown!');
+    } catch (e) {
+      // Test Passed!
+    }
+
+    try {
+      await client.mutate({ mutation, variables });
+      done.fail('Should have thrown!');
+    } catch (e) {
+      // Test Passed!
```

Comment:
@fbartho you can throw the error again here and `expect` it to throw that error. or expect that it resolves to an error.

Reply:
Would you mind, @lifeiscontent, please, giving me an example of how you mean, exactly? I’m new to jest / Typescript, and I had a bear of a time tracing this error through.

`expect(…).rejects` and `expect(…).toThrow()` weren’t clear / didn’t work with the codebase before my change. It was really confusing to see a callstack of an exception thrown when the line that triggered it was in an expect-to-throw block. I felt like what I wrote was marginally cleaner / easier to follow?

Reply:
@fbartho - You could try something like this..

```javascript

    expect.assertions(2);

    await client.query({ query, variables }).catch(
        (e) => expect(e.toString()).toContain('Illegal Query Operation Occurred')
    )

    await client.mutate({ mutation, variables }).catch(
        (e) => expect(e.toString()).toContain('Illegal Mutation Operation Occurred')
    )
      
    done();
```

Reply:
@fbartho you can also do
```js
try {
  await client.mutate({ mutation, variables });
} catch (e) {
  expect(() => {
    throw e;
  }).toThrowErrorMatchingSnapshot();
}
```
I like this way because it saves the error text as a snapshot, so you can easily update the test if you update the error text.

Reply:
@peggyrayzis just pushed an update that matches your syntax. The snapshots raise an interesting point. Errors thrown by `apollo-link-state` resolvers are being treated as ["Network Errors"](https://github.com/apollographql/apollo-link-state/pull/148/commits/4dd7e5b3db52203057ca1ba8af0494cda2fa44a0#diff-445dec7260674092873e1c3f5f39bb51R5)

Is this wrong?

Reply:
@fbartho That's the intended behavior, although the name is a little misleading. Since links are part of the network stack, [any error during the link execution](https://github.com/apollographql/apollo-link/tree/master/packages/apollo-link-error) is treated as a `networkError`. Now that links are being used for more than just network requests, we will probably refine the error API at some point.